### PR TITLE
PoC of python nas heuristic

### PR DESCRIPTION
**Pull request by cooperq**

### CODE REVIEW

File: tools/nasparse.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    if len(sys.argv) != 2:
+        print("usage: nasparse.py [hex encoded nas message]")
+        exit(1)
+        
+    buffer = sys.argv[1]
+    msg = parse_nas_message(buffer)
+    pprint.pprint(msg)
+    res = heur_ue_imsi_sent(msg)
+    if(res[0]): 
+        print(res[1])
```
Comment (wgreenberg) [inline tools/nasparse.py:51]:
```suggestion
    (triggered, message) = heur_ue_imsi_sent(msg)
    if triggered:
        print(message) 
```

File: tools/nasparse.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        parsed = NASLTE.parse_NASLTE_MO(bin)
+    elif uplink == None: #We don't know if its an up or downlink
+        parsed = NASLTE.parse_NASLTE_MO(bin)
+        if parsed[0] == None:
+            parsed = NASLTE.parse_NASLTE_MT(bin)
+    else:
+        parsed = NASLTE.parse_NASLTE_MT(bin)
+    return parsed[0]
+
+def heur_ue_imsi_sent(msg):
```
Comment (wgreenberg) [inline tools/nasparse.py:24]: instead of returning `(bool, string | None)` why not just return `string | None`?

File: tools/nasparse.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        return (False, None)
+
+    if isinstance(msg, pycrate_mobile.TS24301_EMM.EMMSecProtNASMessage):
+        try:
+            msg = msg['EMMAttachRequest']
+        except pycrate_core.elt.EltErr:
+            return (False, None)
+
+    if msg['EPSAttachType']['V'].to_int() == 2:
+        return (True, output)
```
Comment (wgreenberg) [inline tools/nasparse.py:37]: a comment explaining the significance of this would be great

File: tools/nasparse_test.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        return nasparse.heur_ue_imsi_sent(buf)[0]
+
+    def test_imsi_sent(self):
+        self.assertEqual(self.run_heur(self.imsi_sent_msg), True, "imsi_sent_msg should trigger heuristic")
+
+    def test_sec_imsi_sent(self):
+        self.assertEqual(self.run_heur(self.imsi_sent_msg), True, "sec_imsi_sent_msg should trigger heuristic")
+
+    def test_non_nas_msg(self):
+        self.assertEqual(self.run_heur(self.non_nas_msg), False, "non_nas_msg should not trigger heuristic")
```
Comment (wgreenberg) [inline tools/nasparse_test.py:25]: what do you think about throwing an error if we fail to parse a NAS message from the input?

File: tools/nasparse.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    return (False, None)
+
+
+if __name__ == "__main__":
+    if len(sys.argv) != 2:
+        print("usage: nasparse.py [hex encoded nas message]")
+        exit(1)
+        
+    buffer = sys.argv[1]
+    msg = parse_nas_message(buffer)
```
Comment (wgreenberg) [inline tools/nasparse.py:47]: i think we're not handling the case of an invalid NAS message here, in which case i think `msg == None`

**Review by wgreenberg — CHANGES_REQUESTED** [review]

File: tools/nasparse.py (hunk 1 above)
Comment (cooperq) [reply tools/nasparse.py:51]: good suggestion

File: tools/nasparse.py (hunk 2 above)
Comment (cooperq) [reply tools/nasparse.py:24]: I just feel like returning a bool is cleaner especially with the above rewrite

File: tools/nasparse.py (hunk 3 above)
Comment (cooperq) [reply tools/nasparse.py:37]: +1

File: tools/nasparse.py (hunk 5 above)
Comment (cooperq) [reply tools/nasparse.py:47]: it is handled in line 30 by the heuristic, but we could also handle it here instead, never call the heuristic, and also throw an error.

File: tools/nasparse.py [hunk 6]
```diff
@@ -46,7 +51,7 @@ def heur_ue_imsi_sent(msg):
     buffer = sys.argv[1]
     msg = parse_nas_message(buffer)
     pprint.pprint(msg)
-    res = heur_ue_imsi_sent(msg)
-    if(res[0]): 
-        print(res[1])
+    (triggered, message)= heur_ue_imsi_sent(msg)
+    if(triggered):
```
Comment (wgreenberg) [inline tools/nasparse.py:55]: style nit:
```suggestion
    triggered, message = heur_ue_imsi_sent(msg)
    if triggered: 
```

File: tools/pcap_check.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+import sys
+
+TYPE_LTE_NAS = 0x12
+UDP_LEN = 28
+
+def process_pcap(pcap_path):
+    print('Opening {}...'.format(pcap_path))
+
+    count = 0
+    for (pkt_data, pkt_metadata,) in RawPcapNgReader(pcap_path):
```
Comment (wgreenberg) [inline tools/pcap_check.py:13]: nit:
```suggestion
    for pkt_data, pkt_metadata in RawPcapNgReader(pcap_path):
```

File: tools/pcap_check.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+        
+        if gsmtap_hdr[2] != TYPE_LTE_NAS:
+            continue
+
+        # uplink status is the 7th bit of the 5th byte of the GSMTAP header. 
+        # Uplink (Mobile originated) = 0 Downlink (mobile terminated) = 1
+        uplink = (gsmtap_hdr[4] & 0b01000000) >> 6
+        buffer = pkt_data[header_end:]
+        msg = nasparse.parse_nas_message(buffer, uplink)
+        (triggered, message)= nasparse.heur_ue_imsi_sent(msg)
```
Comment (wgreenberg) [inline tools/pcap_check.py:29]: nit:
```suggestion
        triggered, message = nasparse.heur_ue_imsi_sent(msg)
```

File: tools/pcap_check.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    print('Opening {}...'.format(pcap_path))
+
+    count = 0
+    for (pkt_data, pkt_metadata,) in RawPcapNgReader(pcap_path):
+        count += 1
+        gsmtap_len = pkt_data[UDP_LEN+1]  * 4 # gsmtap header length is stored in the 2nd byte of GSMTAP as a number of 32 bit words
+        header_end = gsmtap_len + UDP_LEN #length of UDP/IP header plus GSMTAP header
+        gsmtap_end_idx = (len(pkt_data) - header_end) * -1
+        
+        gsmtap_hdr = pkt_data[UDP_LEN:gsmtap_end_idx]
```
Comment (wgreenberg) [inline tools/pcap_check.py:19]: for some reason this took me forever to wrap my head around lol. would `gsmtap_hdr = pkt_data[UDP_LEN:header_end]` work?

**Review by wgreenberg — CHANGES_REQUESTED** [review]

File: tools/pcap_check.py (hunk 9 above)
Comment (cooperq) [reply tools/pcap_check.py:19]: ah good call, I for some reason forgot you could do a fixed end index for a slice.

**Review by wgreenberg — APPROVED** [review]

**Pull request merged**